src/core/model/Image.cpp

```diff
@@ -45,42 +51,98 @@ void Image::setHeight(double height) {
     this->calcSize();
 }
 
-auto Image::cairoReadFunction(const Image* image, unsigned char* data, unsigned int length) -> cairo_status_t {
-    for (unsigned int i = 0; i < length; i++, image->read++) {
-        if (image->read >= image->data.length()) {
-            return CAIRO_STATUS_READ_ERROR;
-        }
-
-        data[i] = image->data[image->read];
-    }
-
-    return CAIRO_STATUS_SUCCESS;
-}
+void Image::setImage(std::string_view data) { setImage(std::string(data)); }
 
-void Image::setImage(std::string data) {
+void Image::setImage(std::string&& data) {
     if (this->image) {
         cairo_surface_destroy(this->image);
         this->image = nullptr;
     }
     this->data = std::move(data);
+
+    if (this->format) {
+        gdk_pixbuf_format_free(this->format);
+        this->format = nullptr;
+    }
+
+    // FIXME: awful hack to try to parse the format
+    std::array<char*, 4096> buffer{};
+    GdkPixbufLoader* loader = gdk_pixbuf_loader_new();
+    size_t remaining = this->data.size();
+    while (remaining > 0) {
+        size_t readLen = std::min(remaining, buffer.size());
+        if (!gdk_pixbuf_loader_write(loader, reinterpret_cast<const guchar*>(this->data.c_str()), readLen, nullptr))
+            break;
+        remaining -= readLen;
+
+        this->format = gdk_pixbuf_loader_get_format(loader);
+        if (this->format) {
+            // the format is owned by the pixbuf, so create a copy
+            this->format = gdk_pixbuf_format_copy(this->format);
+            break;
+        }
+    }
+    gdk_pixbuf_loader_close(loader, nullptr);
+    g_object_unref(loader);
+
+    g_assert(this->format != nullptr && "could not parse the image format!");
 }
 
-void Image::setImage(GdkPixbuf* img) { setImage(f_pixbuf_to_cairo_surface(img)); }
+void Image::setImage(GdkPixbuf* img) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+    setImage(gdk_cairo_surface_create_from_pixbuf(img, 0, nullptr));
+#pragma GCC diagnostic pop
```

Comment:
Since when gdk_cairo_surface_create_from_pixbuf is deprecated? Also, MSVC does not detect that pragma.